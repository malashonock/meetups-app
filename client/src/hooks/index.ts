export * from './useMeetupQuery';
export * from './useNewsArticleQuery';
export * from './useStaticFileQuery';
export * from './useAuthStore';
