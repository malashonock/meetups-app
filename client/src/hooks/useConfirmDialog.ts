import { useContext } from 'react';

import { ConfirmDialogContext, ConfirmFn } from 'components';
import { Optional } from 'types';

export const useConfirmDialog = (): ConfirmFn => {
  return useContext(ConfirmDialogContext);
};
