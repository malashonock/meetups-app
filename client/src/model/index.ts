export * from './meetup.model';
export * from './news.model';
export * from './user.model';
export * from './login.model';
