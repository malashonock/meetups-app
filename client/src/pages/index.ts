export * from './LoginPage/LoginPage';
export * from './MeetupsPage/MeetupsPage';
export * from './ViewMeetupPage/ViewMeetupPage';
export * from './CreateMeetupPage/CreateMeetupPage';
export * from './EditMeetupPage/EditMeetupPage';
export * from './NewsPage/NewsPage';
export * from './ViewNewsPage/ViewNewsPage';
export * from './CreateNewsPage/CreateNewsPage';
export * from './EditNewsPage/EditNewsPage';
export * from './NotFoundPage/NotFoundPage';
