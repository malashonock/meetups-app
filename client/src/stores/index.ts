export * from './root.store';
export * from './auth.store';
export * from './user.store';
export * from './ui.store';
export * from './meetup.store';
export * from './news.store';
