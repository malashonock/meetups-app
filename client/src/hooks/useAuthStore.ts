import { useContext } from 'react';

import { AuthStore } from 'stores';
import { RootContext } from 'components';

export const useAuthStore = (): AuthStore => {
  return useContext(RootContext).authStore;
};
