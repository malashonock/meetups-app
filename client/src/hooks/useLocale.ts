import { useContext } from 'react';

import { UiStore, Locale } from 'stores';
import { RootContext } from 'components';
import { Optional } from 'types';

export type UseLocaleResult = [
  locale: Optional<Locale>,
  setLocale: (locale: Locale) => void,
];

export const useLocale = (): UseLocaleResult => {
  const uiStore: Optional<UiStore> = useContext(RootContext)?.uiStore;

  const locale = uiStore?.locale;

  const setLocale = (locale: Locale): void => {
    if (uiStore) {
      uiStore.locale = locale;
    }
  };

  return [locale, setLocale];
};
