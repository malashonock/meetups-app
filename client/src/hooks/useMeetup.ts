import { useEffect } from 'react';

import { Meetup } from 'stores';
import { Maybe, Optional } from 'types';
import { useMeetupStore } from 'hooks';

export const useMeetup = (id: Maybe<string>): Optional<Meetup> => {
  const meetupStore = useMeetupStore();

  const meetup = id ? meetupStore.findMeetup(id) : undefined;

  // Hydrate meetup on first load
  useEffect((): void => {
    (async (): Promise<void> => {
      if (meetup && !meetup?.isInitialized) {
        await meetup?.init();
      }
    })();
  }, [meetup, meetup?.isInitialized]);

  return meetup;
};
