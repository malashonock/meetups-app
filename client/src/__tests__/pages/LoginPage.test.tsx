/* eslint-disable testing-library/no-unnecessary-act */

import { PropsWithChildren } from 'react';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import { act, fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { LoginPage } from 'pages';
import { useAuthStore } from 'hooks';
import { AuthStore, RootStore, UserStore } from 'stores';
import { Credentials } from 'model';

// Mock useAuthStore hook
jest.mock('hooks', () => {
  return {
    ...jest.requireActual('hooks'),
    useAuthStore: jest.fn(),
  };
});
const mockUseAuthStore = useAuthStore as jest.MockedFunction<
  typeof useAuthStore
>;
const mockLogIn = jest.fn();

beforeEach(() => {
  mockUseAuthStore.mockReturnValue({
    authStore: {
      rootStore: new RootStore(),
      userStore: new UserStore(new AuthStore(new RootStore())),
      loggedUser: null,
      logIn: mockLogIn,
      logOut: jest.fn(),
      onLoginChanged: jest.fn(),
      toJSON: jest.fn(),
    },
  });
});

afterEach(() => {
  jest.resetAllMocks();
});

const MockRouter = ({ children }: PropsWithChildren): JSX.Element => (
  <MemoryRouter initialEntries={['/', '/login']}>
    <Routes>
      <Route path="/">
        <Route index element={<h1>Home page</h1>} />
        <Route path="login" element={children} />
        <Route path="meetups" element={<h1>Meetups page</h1>} />
      </Route>
    </Routes>
  </MemoryRouter>
);

const getUserNameInput = () =>
  screen.getByLabelText('formFields.login.username.label');
const getPasswordInput = () =>
  screen.getByLabelText('formFields.login.password.label');
const getLoginBtn = () => screen.getByText('formButtons.login');

const mockCredentials: Credentials = {
  username: 'John',
  password: 'sweethomealabama',
};

describe('LoginPage', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(<LoginPage />, { wrapper: MockRouter });

    expect(asFragment()).toMatchSnapshot();
  });

  it('should open a blank form', () => {
    render(<LoginPage />, { wrapper: MockRouter });

    expect(getUserNameInput()).toHaveValue('');
    expect(getPasswordInput()).toHaveValue('');
  });

  it('should accept user input', async () => {
    render(<LoginPage />, { wrapper: MockRouter });

    await act(() => {
      userEvent.type(getUserNameInput(), mockCredentials.username);
    });
    expect(getUserNameInput()).toHaveValue(mockCredentials.username);

    await act(() => {
      userEvent.type(getPasswordInput(), mockCredentials.password);
    });
    expect(getPasswordInput()).toHaveValue(mockCredentials.password);
  });

  it('should validate form fields', async () => {
    render(<LoginPage />, { wrapper: MockRouter });

    expect(getLoginBtn()).toBeDisabled();

    await act(() => {
      fireEvent.blur(getUserNameInput());
    });
    expect(
      screen.getByText('formFields.login.username.errorText'),
    ).toBeInTheDocument();

    await act(() => {
      fireEvent.blur(getPasswordInput());
    });
    expect(
      screen.getByText('formFields.login.password.errorText'),
    ).toBeInTheDocument();
  });

  it('should handle form submit', async () => {
    render(<LoginPage />, { wrapper: MockRouter });

    await act(() => {
      userEvent.type(getUserNameInput(), mockCredentials.username);
      userEvent.type(getPasswordInput(), mockCredentials.password);
    });

    await act(() => {
      userEvent.click(getLoginBtn());
    });

    expect(mockLogIn).toHaveBeenCalledWith(mockCredentials);
    expect(screen.getByText('Meetups page')).toBeInTheDocument();
  });
});
