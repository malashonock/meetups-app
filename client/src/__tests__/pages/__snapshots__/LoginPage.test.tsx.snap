// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage should match snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <section
      class="container"
    >
      <h1
        class="typography h1 heading"
      >
        loginPage.title
      </h1>
      <div
        class="contentWrapper"
      >
        <div
          class="textSection main"
        >
          <figure
            class="section imageWrapper"
          >
            <svg
              class="image"
            >
              anonymous-user.svg
            </svg>
          </figure>
          <div
            class="container"
          >
            <label
              class="inputLabel label"
              for="username"
            >
              formFields.login.username.label
            </label>
            <input
              class="input default"
              data-testid="text-input"
              id="username"
              name="username"
              type="text"
              value=""
            />
          </div>
          <div
            class="container"
          >
            <label
              class="inputLabel label"
              for="password"
            >
              formFields.login.password.label
            </label>
            <input
              class="input default"
              data-testid="text-input"
              id="password"
              name="password"
              type="text"
              value=""
            />
          </div>
          <button
            class="actionButton button primary"
            disabled=""
            id="btn-login"
            type="submit"
          >
            formButtons.login
          </button>
        </div>
      </div>
    </section>
  </form>
</DocumentFragment>
`;
